storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3110,1754 @@ buf_block_t *fsp_get_latched_xdes_page(
+    }
+
+    buf_block_t *header_block=
+      fsp_get_latched_page(page_id_t{0, 0}, mtr, &err);
+    if (!header_block)
+      return err;
+
+    const fseg_header_t *seg_header= m_root->page.frame +
+      (level ? PAGE_HEADER + PAGE_BTR_SEG_TOP
+             : PAGE_HEADER + PAGE_BTR_SEG_LEAF);
+
+    buf_block_t *iblock;
+    fseg_inode_t *inode= fseg_inode_try_get(seg_header, 0, 0, mtr,
+                                            &iblock, &err);
+    if (!inode)
+      return err;
+
+    auto parent_it= m_parent_pages.find(cur_page_no);
+    if (parent_it == m_parent_pages.end())
+    {
+      err= DB_CORRUPTION;
+      return err;
+    }
+
+    uint32_t parent_page_no= uint32_t(parent_it->second);
+
+    uint32_t parent_offset= uint32_t(parent_it->second >> 32);
+
+    if (parent_offset > srv_page_size - FIL_PAGE_DATA_END)
+    {
+      err= DB_CORRUPTION;
+      return err;
+    }
+
+    pageOperator operation(header_block, iblock, inode, extent_size,
+                           cur_page_no, mtr);
+
+    associatedPages related_pages(block, mtr);
+    err= operation.prepare_new_page(new_extent, old_state == XDES_FSEG);
+
+    DBUG_EXECUTE_IF("allocation_prepare_fail", err= DB_CORRUPTION;);
+    if (err) return err;
+
+    err= related_pages.prepare(parent_page_no);
+    DBUG_EXECUTE_IF("relation_page_prepare_fail", err= DB_CORRUPTION;);
+
+    if (err) return err;
+
+    operation.complete_new_page();
+
+    /* After allocating the new page, try to prepare the steps
+    of page removal function. Because there is a possiblity that
+    last block in FSEG_NOT_FULL/FSP_FREE_FRAG/FSP_FREE last block
+    could've changed while allocating the new block. */
+    err= operation.prepare_old_page();
+
+    DBUG_EXECUTE_IF("remover_prepare_fail", err= DB_CORRUPTION;);
+    if (err)
+    {
+      operation.restore_old_pages();
+      mtr->discard_modifications();
+      return err;
+    }
+
+    /* Copy the data from old block to new block */
+    buf_block_t *new_block= operation.get_new_block();
+    uint32_t new_page_no= new_block->page.id().page_no();
+    /* Copy FIL_PAGE_PREV, FIL_PAGE_NEXT */
+    mtr->memcpy<mtr_t::MAYBE_NOP>(*new_block,
+                                  new_block->page.frame + FIL_PAGE_PREV,
+                                  block->page.frame + FIL_PAGE_PREV,
+                                  page_has_next(block->page.frame) ? 8 : 4);
+    mtr->memcpy(*new_block, new_block->page.frame + FIL_PAGE_TYPE,
+                block->page.frame + FIL_PAGE_TYPE,
+                srv_page_size - FIL_PAGE_TYPE - 8);
+
+    /* Assign the new block page number in left, right
+    and parent block */
+    related_pages.complete(new_page_no, parent_offset);
+
+    /* Complete the page free operation */
+    operation.complete_free_old_page();
+    /* Add the new page in inode fragment array */
+    operation.assign_frag_slot();
+
+    if (level)
+    {
+      err= store_child_pages(new_block);
+      if (err) return err;
+    }
```

Comment:
We seem to be missing a `prepare` step for this one, or we are missing `operation.restore_old_pages()` and `mtr->discard_modifications()` if this would fail.